opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -2485,143 +2485,258 @@ bool CpGrid::adapt(const std::vector<std::array<int,3>>& cells_per_dim_vec,
-        assert(static_cast<std::size_t>(current_data_->back()->cellIndexSet().size()) == static_cast<std::size_t>(current_data_->back()->size(0)) );
+    // Compute the partition type for cell
+    (*current_data_).back()->computeCellPartitionType();
+
+    // Compute the partition type for point
+    (*current_data_).back()->computePointPartitionType();
+
+    // Now we can compute the communication interface.
+    current_data_->back()->computeCommunicationInterfaces(current_data_->back()->size(3));
+    assert(static_cast<std::size_t>(current_data_->back()->cellIndexSet().size()) == static_cast<std::size_t>(current_data_->back()->size(0)) );
 #endif
-    } // end-if-comm().size()>1
+}
 
+void CpGrid::getFirstChildGlobalIds([[maybe_unused]] std::vector<int>& parentToFirstChildGlobalIds)
+{
+#if HAVE_MPI
+    switchToGlobalView();
 
-    // Print total amount of cells on the adapted grid
-    Opm::OpmLog::info(std::to_string(markedElem_count) + " elements have been marked (in " + std::to_string(comm().rank()) + " rank).\n");
-    Opm::OpmLog::info(std::to_string(levels)  + " (new) refined level grid(s) (in " + std::to_string(comm().rank()) + " rank).\n");
-    Opm::OpmLog::info(std::to_string(cell_count)  + " total cells on the leaf grid view (in " + std::to_string(comm().rank()) + " rank).\n");
+    const auto& data = currentData();
+    const auto& parentToChildrenBeforeLoadBalance = data[0]->getParentToChildren();
+    const auto& globalIdSet = this->globalIdSet();
 
-    return preAdapt();
+    parentToFirstChildGlobalIds.resize(data[0]->size(0), -1); // Initialize with -1 (invalid value, for non parent cells).
+
+    const auto& elements = Dune::elements(levelGridView(0));
+    for (const auto& element : elements) {
+        const auto& [level, children] = parentToChildrenBeforeLoadBalance[element.index()];
+
+        if (!children.empty()) {
+            const auto& levelData = *data[level];
+            const auto& first_child = Dune::cpgrid::Entity<0>(levelData, children[0], true);
+
+            // Rewrite parent global id entry with first child global id
+            parentToFirstChildGlobalIds[globalIdSet.id(element)] = globalIdSet.id(first_child);
+        }
+    }
+#endif
 }
 
-void CpGrid::postAdapt()
+void CpGrid::syncCellIds()
 {
-    // - Resize with the new amount of cells on the leaf grid view
-    // - Set marks equal to zero (representing 'doing nothing')
-    current_view_data_ -> postAdapt();
+#if HAVE_MPI
+    std::vector<int> parentToFirstChildGlobalIds;
+    getFirstChildGlobalIds(parentToFirstChildGlobalIds);
+
+    auto size = comm().max(parentToFirstChildGlobalIds.size());
+
+    switchToDistributedView();
+
+    std::vector<int> firstChildGlobalIds(size);
+    if (comm().rank() == 0) {
+        firstChildGlobalIds = parentToFirstChildGlobalIds;
+    }
+    comm().broadcast(firstChildGlobalIds.data(), firstChildGlobalIds.size(), 0);
```

Comment:
Can't we omit this container just use parentToFirstChildGlobalIds directly?

Reply:
Thanks for the improvement. parentToFristChildGlobalIds does the job too

Reply:
This was for the next iteration. Please ignore.